### PR TITLE
ATO-146: Create identity progress handler

### PR DESCRIPTION
**Pull request by CarlyG55**
## What?

This PR creates the identity progress handler for the orchestration frontend to use to check if SPOT has returned identity credentials whilst the spinner page is showing. This PR create the `BaseOrchestrationHandler` that AuthCodeHandler will extend in the future once this lambda is under the remit of Orchestration. There is also some minor refactoring to move `AuthenticationUserInfo` and `AuthenticationUserInfoStorageService` to the shared folder. This move is implemented in the first commit for ease of reviewing.

## Why?

The existing `ProcessingIdentityHandler` requires the email to be posted, however the Orchestration frontend does not hold state, so does not have this.

## Related PRs
https://github.com/govuk-one-login/orchestration-frontend/pull/8

### CODE REVIEW

File: ipv-api/src/main/java/uk/gov/di/authentication/ipv/entity/IdentityProgressResponse.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+import java.net.URI;
+
+public class IdentityProgressResponse {
+
+    @SerializedName("status")
+    @Expose
+    @Required
+    private IdentityProgressStatus status;
+
+    @SerializedName("client-name")
```
Comment (CarlyG55) [inline ipv-api/src/main/java/uk/gov/di/authentication/ipv/entity/IdentityProgressResponse.java:17]: There doesn't seem to be a consistent casing for serialized name. I've chosen the same casing as AuthCodeHandler, as that's the other orchestration handler we'll have

File: ipv-api/src/main/java/uk/gov/di/authentication/ipv/entity/IdentityProgressResponse.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    @Expose
+    @Required
+    private String clientName;
+
+    @SerializedName("redirect-uri")
+    @Expose
+    @Required
+    private URI redirectUri;
```
Comment (CarlyG55) [inline ipv-api/src/main/java/uk/gov/di/authentication/ipv/entity/IdentityProgressResponse.java:25]: Similarly, there isn't consistency between if we should return uri/state as strings or their given types - I'm happy to return strings if people think that's the correct option

File: shared/src/main/java/uk/gov/di/authentication/shared/lambda/BaseOrchestrationHandler.java
Comment (ethanmills) [inline shared/src/main/java/uk/gov/di/authentication/shared/lambda/BaseOrchestrationHandler.java:1]: Is this a BaseOrchestrationHandler or a BaseOrchestationFrontendHandler?

File: shared/src/main/java/uk/gov/di/authentication/shared/lambda/BaseOrchestrationHandler.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        this.configurationService = configurationService;
+        this.sessionService = new SessionService(configurationService);
+        this.clientSessionService = new ClientSessionService(configurationService);
+    }
+
+    @Override
+    public APIGatewayProxyResponseEvent handleRequest(
+            APIGatewayProxyRequestEvent input, Context context) {
+        return segmentedFunctionCall(
+                "ipv-api::" + getClass().getSimpleName(),
```
Comment (ethanmills) [inline shared/src/main/java/uk/gov/di/authentication/shared/lambda/BaseOrchestrationHandler.java:56]: Is this always the right segment?

File: shared/src/main/java/uk/gov/di/authentication/shared/lambda/BaseOrchestrationHandler.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+        userSessionBuilder.withClientSession(clientSession.orElse(null));
+
+        var clientID =
+                clientSession
+                        .map(ClientSession::getAuthRequestParams)
+                        .map(t -> t.get(CLIENT_ID))
+                        .flatMap(v -> v.stream().findFirst());
+        attachLogFieldToLogs(LogLineHelper.LogFieldName.CLIENT_ID, clientID.orElse(UNKNOWN));
+        userSessionBuilder.withClientId(clientID.orElse(null));
+
```
Comment (ethanmills) [inline shared/src/main/java/uk/gov/di/authentication/shared/lambda/BaseOrchestrationHandler.java:95]: Would be good to have the persistent session ID stuff here too, but not fail if it's not present

File: shared/src/main/java/uk/gov/di/authentication/shared/state/UserSession.java
Comment (ethanmills) [inline shared/src/main/java/uk/gov/di/authentication/shared/state/UserSession.java:1]: Might be worth calling this OrchestrationUserSession for now to make it clear who's concern it is

File: ipv-api/src/main/java/uk/gov/di/authentication/ipv/entity/IdentityProgressResponse.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+    @SerializedName("state")
+    @Expose
+    @Required
+    private State state;
+
+    public IdentityProgressResponse() {}
```
Comment (ethanmills) [inline ipv-api/src/main/java/uk/gov/di/authentication/ipv/entity/IdentityProgressResponse.java:32]: Why do we need the no-args constructor here?

File: ipv-api/src/main/java/uk/gov/di/authentication/ipv/entity/IdentityProgressResponse.java
Comment (ethanmills) [inline ipv-api/src/main/java/uk/gov/di/authentication/ipv/entity/IdentityProgressResponse.java:1]: Could this be a record, or would that break the serialisation?

File: ipv-api/src/main/java/uk/gov/di/authentication/ipv/entity/IdentityProgressResponse.java (hunk 2 above)
Comment (ethanmills) [reply ipv-api/src/main/java/uk/gov/di/authentication/ipv/entity/IdentityProgressResponse.java:25]: Happy with this. An approval test is a good way to check the serialisation if that's something you're concerned about

File: shared/src/main/java/uk/gov/di/authentication/shared/lambda/BaseOrchestrationHandler.java
Comment (CarlyG55) [reply shared/src/main/java/uk/gov/di/authentication/shared/lambda/BaseOrchestrationHandler.java:1]: I wasn't really sure if `AuthCodeHandler` is considered frontend or not

File: shared/src/main/java/uk/gov/di/authentication/shared/lambda/BaseOrchestrationHandler.java (hunk 3 above)
Comment (CarlyG55) [reply shared/src/main/java/uk/gov/di/authentication/shared/lambda/BaseOrchestrationHandler.java:56]: Yeah, this is one I was a little unsure of. I guess it's going to depend on where this is implemented, and will change when we move to our own repo. I'm not really sure what we should replace if with though

File: shared/src/main/java/uk/gov/di/authentication/shared/lambda/BaseOrchestrationHandler.java (hunk 4 above)
Comment (CarlyG55) [reply shared/src/main/java/uk/gov/di/authentication/shared/lambda/BaseOrchestrationHandler.java:95]: I think elsewhere with the persistent session we just add to logs - is that what you mean?

File: ipv-api/src/main/java/uk/gov/di/authentication/ipv/entity/IdentityProgressResponse.java (hunk 5 above)
Comment (CarlyG55) [reply ipv-api/src/main/java/uk/gov/di/authentication/ipv/entity/IdentityProgressResponse.java:32]: I think this is just me not having that much java experience, and having seen other examples in the code so thinking I needed it 😅

File: ipv-api/src/main/java/uk/gov/di/authentication/ipv/entity/IdentityProgressResponse.java
Comment (CarlyG55) [reply ipv-api/src/main/java/uk/gov/di/authentication/ipv/entity/IdentityProgressResponse.java:1]: I *think* it should be alright https://github.com/google/gson/releases/tag/gson-parent-2.10

File: ipv-api/src/main/java/uk/gov/di/authentication/ipv/entity/IdentityProgressResponse.java (hunk 2 above)
Comment (CarlyG55) [reply ipv-api/src/main/java/uk/gov/di/authentication/ipv/entity/IdentityProgressResponse.java:25]: Added one to check the above as well

**Review by ethanmills — APPROVED** [review]

**Pull request merged**